### PR TITLE
Fixes #20895: Corrected handling of XYWH boxes.

### LINKED ISSUE #20895: Incorrect processing of XYWH boxes in `scale_boxes`
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Other

### Bug

`scale_boxes` function's docstring says that it supports both `xyxy` and `xywh` box formats, but its logic for `xywh` format is incorrect.

`scale_boxes` calls `clip_boxes` towards the end:

https://github.com/ultralytics/ultralytics/blob/7564930f1ad841ab9972712c5508c47c89480e54/ultralytics/utils/ops.py#L140

but `clip_boxes` does not handles `xywh` format. Therefore, `scale_boxes` can return boxes like `[110.,  30., 180., 180.]` (see example), which run out of the image.

### Environment

```
Ultralytics 8.3.146 🚀 Python-3.11.12 torch-2.6.0+cu124 CPU (Intel Xeon 2.20GHz)
Setup complete ✅ (2 CPUs, 12.7 GB RAM, 41.4/107.7 GB disk)

OS                  Linux-6.1.123+-x86_64-with-glibc2.35
Environment         Colab
Python              3.11.12
Install             pip
Path                /usr/local/lib/python3.11/dist-packages/ultralytics
RAM                 12.67 GB
Disk                41.4/107.7 GB
CPU                 Intel Xeon 2.20GHz
CPU count           2
GPU                 None
GPU count           None
CUDA                None

numpy               ✅ 2.0.2>=1.23.0
matplotlib          ✅ 3.10.0>=3.3.0
opencv-python       ✅ 4.11.0.86>=4.6.0
pillow              ✅ 11.2.1>=7.1.2
pyyaml              ✅ 6.0.2>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.15.3>=1.4.1
torch               ✅ 2.6.0+cu124>=1.8.0
torch               ✅ 2.6.0+cu124!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.21.0+cu124>=0.9.0
tqdm                ✅ 4.67.1>=4.64.0
psutil              ✅ 5.9.5
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.2>=1.1.4
ultralytics-thop    ✅ 2.0.14>=2.0.0
{'OS': 'Linux-6.1.123+-x86_64-with-glibc2.35',
 'Environment': 'Colab',
 'Python': '3.11.12',
 'Install': 'pip',
 'Path': '/usr/local/lib/python3.11/dist-packages/ultralytics',
 'RAM': '12.67 GB',
 'Disk': '41.4/107.7 GB',
 'CPU': 'Intel Xeon 2.20GHz',
 'CPU count': 2,
 'GPU': None,
 'GPU count': None,
 'CUDA': None,
 'Package Info': {'numpy': '✅ 2.0.2>=1.23.0',
  'matplotlib': '✅ 3.10.0>=3.3.0',
  'opencv-python': '✅ 4.11.0.86>=4.6.0',
  'pillow': '✅ 11.2.1>=7.1.2',
  'pyyaml': '✅ 6.0.2>=5.3.1',
  'requests': '✅ 2.32.3>=2.23.0',
  'scipy': '✅ 1.15.3>=1.4.1',
  'torch': '✅ 2.6.0+cu124!=2.4.0,>=1.8.0; sys_platform == "win32"',
  'torchvision': '✅ 0.21.0+cu124>=0.9.0',
  'tqdm': '✅ 4.67.1>=4.64.0',
  'psutil': '✅ 5.9.5',
  'py-cpuinfo': '✅ 9.0.0',
  'pandas': '✅ 2.2.2>=1.1.4',
  'ultralytics-thop': '✅ 2.0.14>=2.0.0'}}
```

### Minimal Reproducible Example

```python
import torch
from ultralytics.utils.ops import scale_boxes, xyxy2xywh, xywh2xyxy

img1_shape = (320, 320)  # shape of processed image
img0_shape = (480, 640)  # shape of original image

boxes = torch.tensor([[10, 10, 100, 100]], dtype=torch.float32)  # xyxy

scaled_xyxy = scale_boxes(img1_shape=img1_shape, boxes=boxes.clone(), img0_shape=img0_shape, xywh=False)
print("Scaled XYXY:", scaled_xyxy)

boxes_xywh = xyxy2xywh(boxes)
scaled_xywh = scale_boxes(img1_shape=img1_shape, boxes=boxes_xywh.clone(), img0_shape=img0_shape, xywh=True)
print("Scaled XYWH:", scaled_xywh)
print("Scaled XYWH -> XYXY:", xywh2xyxy(scaled_xywh))
```

Output:
```
Scaled XYXY: tensor([[ 20.,   0., 200., 120.]])
Scaled XYWH: tensor([[110.,  30., 180., 180.]])
Scaled XYWH -> XYXY: tensor([[ 20., -60., 200., 120.]])
```

### Additional

_No response_

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by anmolojha**
Fixes #20895.

`scale_boxes` now converts `xywh` boxes to `xyxy` format before processing.

### Verification
Output of Minimal Reproducible Example provided in the issue:

Before:
```
Scaled XYXY: tensor([[ 20.,   0., 200., 120.]])
Scaled XYWH: tensor([[110.,  30., 180., 180.]])
Scaled XYWH -> XYXY: tensor([[ 20., -60., 200., 120.]])
```

Now:
```
Scaled XYXY: tensor([[ 20.,   0., 200., 120.]])
Scaled XYWH: tensor([[110.,  60., 180., 120.]])
Scaled XYWH -> XYXY: tensor([[ 20.,   0., 200., 120.]])
```

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Adds proper support for oriented bounding boxes (OBB, xywhr) and fixes clipping behavior for xywh boxes in `scale_boxes`, improving accuracy and consistency when rescaling boxes between image sizes. 🧭📦

### 📊 Key Changes
- `scale_boxes` now supports xyxy, xywh, and xywhr (OBB) formats.
- Updated docstrings to clearly document OBB handling and expected formats.
- Ensures correct clipping by:
  - Converting xywh → xyxy before clipping, then converting back to xywh.
  - Preserving OBB (xywhr) without unnecessary format conversions.
- Maintains angle (r) handling for OBBs while scaling and padding adjustments apply to x, y, w, h.

### 🎯 Purpose & Impact
- More reliable post-processing for OBB models (e.g., YOLO11 OBB variants) when resizing or letterboxing images. 🧰
- Fixes potential mis-clipping of xywh boxes, preventing distorted or out-of-bounds boxes. ✅
- Backward compatible for existing xyxy/xywh workflows—no user changes required.
- Leads to more accurate visualization and evaluation when moving predictions between image shapes, benefiting training, inference, and Ultralytics HUB integrations. 🚀

Example:
```python
from ultralytics.utils.ops import scale_boxes
# boxes_xywhr: tensor[N, 5] -> [x, y, w, h, r]
boxes_scaled = scale_boxes(img1_shape, boxes_xywhr, img0_shape, xywh=True)
```

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @anmolojha, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! This is an automated response to help streamline the review process. An Ultralytics engineer will review your contribution and assist further soon.

To ensure a smooth integration, please review and confirm the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

Since this PR appears to address a bug, if you haven't already, please include a **minimum reproducible example (MRE)** in the PR description to help us quickly verify and test your fix.

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don’t hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀🦾

**Comment by anmolojha** [comment]
I have read the CLA Document and I sign the CLA

**Comment by anmolojha** [comment]
@glenn-jocher views on failing tests?

This turned out to be a more difficult problem than I previously thought 😅.

Unless I am missing something, converting bounding boxes from letterboxed to original image is not trivial for oriented bounding boxes. See following figures:

Screenshot 2025-06-07 at 9 32 48 PM
<p align="center">Fig 1</p>


Screenshot 2025-06-07 at 10 05 23 PM
<p align="center">Fig 2</p>

Am I making any mistake? Otherwise, are there any thoughts on how to approach this?

**Comment by glenn-jocher** [comment]
@anmolojha You're absolutely right that oriented bounding box scaling is more complex! The issue is that OBB center coordinates need special handling during the padding removal step.

For oriented boxes in `xywhr` format, the current approach of subtracting padding from all coordinates doesn't work correctly because only the center coordinates (x, y) should have padding subtracted, while width, height, and rotation should remain unchanged during the padding removal step.

Consider modifying the padding logic to handle OBB specifically - you'll want to check if the boxes have 5 coordinates (indicating `xywhr` format) and only apply padding adjustments to the center coordinates `[..., :2]` rather than all coordinates `[..., :4]`. The [reference documentation for OBB conversions](https://docs.ultralytics.com/reference/utils/ops/) shows functions like `xywhr2xyxyxyxy` that might be helpful for validation.

Have you tried isolating the padding removal logic for the 5-coordinate case?

**Comment by anmolojha** [comment]
@glenn-jocher, what you’re suggesting works for OBBs that are well within the image boundaries. However, for OBBs near the image edge, it's not straightforward to map them back to the original image — see Fig 2 in the [previous comment](https://github.com/ultralytics/ultralytics/pull/20971#issuecomment-2952739630). The challenge is that **once padding is removed, the OBB may no longer form a quadrilateral.**

The current implementation in the repo does what you mentioned (subtracting padding from the center coordinates `c_x` and `c_y`), but this doesn't ensure non-negative coordinates. See the code below:

```python
import math
import torch
from ultralytics.utils.ops import scale_boxes, xywhr2xyxyxyxy

img1_shape = (320, 320)  # shape of processed image
img0_shape = (320, 200)  # shape of original image

boxes = torch.tensor([[100, 100, 100, 100, math.pi / 4]], dtype=torch.float32)  # xywhr
boxes_xyxyxyxy = xywhr2xyxyxyxy(boxes)

print("Initial Boxes (XYWHR):", boxes)
print("Initial Boxes (XYXYXYXY):", boxes_xyxyxyxy)

scaled_xywhr = scale_boxes(img1_shape=img1_shape, boxes=boxes.clone(), img0_shape=img0_shape, xywh=True)
print("Scaled Boxes (XYWHR):", scaled_xywhr)
print("Scaled Boxes (XYXYXYXY):", xywhr2xyxyxyxy(scaled_xywhr))
```

Output:
```
Initial Boxes (XYWHR): tensor([[100.0000, 100.0000, 100.0000, 100.0000,   0.7854]])
Initial Boxes (XYXYXYXY): tensor([[[100.0000, 170.7107],
         [170.7107, 100.0000],
         [100.0000,  29.2893],
         [ 29.2893, 100.0000]]])
Scaled Boxes (XYWHR): tensor([[ 40.0000, 100.0000, 100.0000, 100.0000,   0.7854]])
Scaled Boxes (XYXYXYXY): tensor([[[ 40.0000, 170.7107],
         [110.7107, 100.0000],
         [ 40.0000,  29.2893],
         [-30.7107, 100.0000]]])
```

**Comment by glenn-jocher** [comment]
@anmolojha You've done excellent analysis identifying this geometric constraint! You're absolutely correct that naive padding removal for OBBs can produce invalid coordinates when boxes extend beyond boundaries.

This reveals that `scale_boxes` needs a more sophisticated approach for OBBs. A few potential solutions:

1. **Clip to valid region**: Convert to `xyxyxyxy`, clip coordinates to `[0, image_width/height]`, then convert back (though this changes the OBB shape)

2. **Skip OBB scaling**: Since OBBs have rotation, simple coordinate scaling may not preserve the intended object geometry anyway - consider whether OBBs should use this function at all

3. **Transform then validate**: Apply the current logic but add validation to ensure all corner coordinates remain valid, with fallback handling for edge cases

Given the geometric complexity you've uncovered, I'd lean toward option 2 - OBBs might need specialized scaling logic rather than using the general `scale_boxes` function. What's your assessment of how critical OBB support is for this fix versus addressing the original XYWH issue first?

**Comment by anmolojha** [comment]
Hi,
Yes, I also think option 2 is the best. It correctly handles scaling for `xywh` format and leaves `xywhr` unchanged. Handling OBBs near the image boundary is a complex issue and deserves more careful consideration.

However, I don't think this is correct:
> simple coordinate scaling may not preserve the intended object geometry anyway

Scaling is fine—it preserves rotation angle as long as the aspect ratio is maintained, which is the case with `LetterBox`. If the aspect ratio changes (e.g., stretching the image horizontally or vertically), the angles will change—tending toward `0` or `π/2`, respectively.

For now, OBBs should behave correctly in most cases. In edge cases where they fall near the boundary, the current behaviour remains unchanged—returning negative coordinates. I was thinking of adding a warning when that happens. Let me know what you think.

**Comment by Laughing-q** [comment]
@anmolojha Hey thanks for the PR, I think this way would complicate things a bit if passing `xywh=True`, as in this PR it'd check if it's obb, and convert to `xyxy` if it's not obb and then do clip_boxes then convert back to xywh. 
The first intention we introduced `xywh` option here is to maximumly reuse existing code instead of creating another function that has duplicated code, so it's kinda not designed to cover the best practice for `xywh` format, but to make it compatible with `xywh` format.
I would recommend you to do a clip_boxes after getting results from `scale_boxes(xywh=True)`.

**Comment by anmolojha** [comment]
Hi @Laughing-q,
Yes, I agree this PR complicates the code a bit. The better way would be to make `clip_boxes` more robust to box formats, but as I mentioned in the [original issue](https://github.com/ultralytics/ultralytics/issues/20895), modifying it by adding an argument `xywh=True` feels like a patchy solution. I'd prefer to keep patchy logic within already patchy code rather than introduce it into cleaner parts.

> I would recommend you to do a clip_boxes after getting results from scale_boxes(xywh=True)

You’ll have to explain a bit more why you suggest separating them. I am thinking of `scale_boxes` as a one-stop solution for converting boxes between images—it handles scaling and padding—so I’d expect it to return valid (clipped) output. I also can't think of a good place where you would call only `scale_boxes` and not `clip_boxes`.
